explorer/explorer_android.go

```diff
@@ -157,109 +231,52 @@ func (f *File) Write(b []byte) (n int, err error) {
 	return n, err
 }
 
-func (f *File) Close() error {
+func (f *FileWriter) Close() error {
 	if f == nil || f.isClosed {
 		return io.ErrClosedPipe
 	}
 	f.isClosed = true
 
 	return jni.Do(jni.JVMFor(app.JavaVM()), func(env jni.Env) error {
-		mutex.Lock()
-		defer mutex.Unlock()
-
-		defer func() {
-			if f.obj != 0 {
-				jni.DeleteGlobalRef(env, f.obj)
-			}
-			if f.sharedBuffer != 0 {
-				jni.DeleteGlobalRef(env, f.sharedBuffer)
-			}
-		}()
-
-		flush := jni.Value(0)
-		if f.isWritable {
-			flush = jni.Value(f.obj)
-		}
-
-		ok, err := jni.CallBooleanMethod(env, _LibObject, _FileClose, jni.Value(f.obj), flush)
+		ok, err := jni.CallBooleanMethod(env, f.libObject, f.fileClose, jni.Value(f.obj), jni.Value(f.obj))
 		if err != nil {
 			return err
 		}
 
+		jni.DeleteGlobalRef(env, f.obj)
+
 		if !ok {
 			return io.ErrClosedPipe
 		}
 		return nil
 	})
 }
 
-func openFile(extensions ...string) (io.ReadCloser, error) {
-	for i, ext := range extensions {
-		extensions[i] = mime.TypeByExtension(ext)
-	}
-
-	err := jni.Do(jni.JVMFor(app.JavaVM()), func(env jni.Env) error {
-		mutex.Lock()
-		defer mutex.Unlock()
-
-		return jni.CallVoidMethod(env, _LibObject, _OpenFile,
-			jni.Value(view),
-			jni.Value(jni.JavaString(env, strings.Join(extensions, ","))),
-		)
-	})
-
-	if err != nil {
-		return nil, err
-	}
-
-	file := <-openFileCallback
-	if file == nil {
-		return nil, ErrUserDecline
-	}
-	return file, nil
-}
-
-func createFile(name string) (io.WriteCloser, error) {
-	err := jni.Do(jni.JVMFor(app.JavaVM()), func(env jni.Env) error {
-		mutex.Lock()
-		defer mutex.Unlock()
-
-		return jni.CallVoidMethod(env, _LibObject, _CreateFile,
-			jni.Value(view),
-			jni.Value(jni.JavaString(env, strings.TrimPrefix(strings.ToLower(filepath.Ext(name)), "."))),
-		)
-	})
-	if err != nil {
-		return nil, err
-	}
-
-	file := <-createFileCallback
-	if file == nil {
-		return nil, ErrUserDecline
-	}
-	return file, nil
-}
-
-//export Java_org_gioui_x_explorer_explorer_1android_OpenCallback
-func Java_org_gioui_x_explorer_explorer_1android_OpenCallback(env *C.JNIEnv, _ C.jclass, b C.jobject) {
-	if b == 0 {
-		openFileCallback <- nil
-	} else {
-		openFileCallback <- &File{
-			obj:        jni.NewGlobalRef(jni.EnvFor(uintptr(unsafe.Pointer(env))), jni.Object(uintptr(b))),
-			isReadable: true,
+//export Java_org_gioui_x_explorer_explorer_1android_ImportCallback
+func Java_org_gioui_x_explorer_explorer_1android_ImportCallback(env *C.JNIEnv, _ C.jclass, b C.jobject, id C.jint) {
+	if v, ok := active.Load(int32(id)); ok {
+		v := v.(*explorer)
+		if b == 0 {
+			v.result <- result{error: ErrUserDecline}
+		} else {
+			v.result <- result{file: newFileReader(v, jni.NewGlobalRef(jni.EnvFor(uintptr(unsafe.Pointer(env))), jni.Object(uintptr(b))))}
```

Comment:
That `jni.NewGlobalRef` (also on `ExportCallback`) was never released. It was fix in the latest patch. ;)